allennlp/semparse/worlds/atis_world.py

```diff
@@ -31,74 +59,182 @@ def __init__(self,
                  utterances: List[str],
                  tokenizer: Tokenizer = None,
                  database_directory: str = None) -> None:
+
+        self.all_tables = ALL_TABLES
+        self.tables_with_strings = TABLES_WITH_STRINGS
         if database_directory:
-            self.sql_table_context = SqlTableContext(ALL_TABLES,
-                                                     TABLES_WITH_STRINGS,
-                                                     database_directory)
+            self.database_directory = database_directory
+            self.connection = sqlite3.connect(self.database_directory)
+            self.cursor = self.connection.cursor()
+
+        self.sql_table_context = SqlTableContext(ALL_TABLES,
```

Comment:
Does the order of this mean that you have two concurrent connections to the db open?  I'm not sure what performance issues / race conditions that might induce.  It's probably fine.

Reply:
We are only doing reads here, so multiple processes can access the database without any locking/waiting (https://www.sqlite.org/faq.html#q5).